lib/guardian/plug/pipeline.ex

```diff
@@ -0,0 +1,192 @@
+  refres to, failing if one is not found.
+
+  ### Customizing your pipeline
+
+  Once you've created a pipeline, you can customize it when you call it with options.
+
+  ```elixir
+  plug MyApp.AuthPipeline, module: MyApp.ADifferentGuardianModule
+  # OR
+  plug MyApp.AuthPipeline, key: :impersonate
+  ```
+
+  ### Options
+
+  You can provide options to the pipeline when you `use Guardian.Plug.Pipeline`
+  or you can provide them when you call the plug.
+
+  Additionally, for every option other than `:otp_app` you can use elixir
+  configuration, the using options, or linline options.
+
+  * `:otp_app` - The otp app where the pipeline modules can be found
+  * `:module` - The `Guardian` implementation module
+  * `:error_handler` - An error handling module. See `Guardian.Plug.Errors`
+  * `:key` - The key to use
+
+  ### Keys
+
+  Using keys allows you to specifiy locations in the session/connection where
+  the tokens and resources will be placed. This allows multiple authenticated
+  tokens to be in play for a single request. This is useful for impersonation or
+  higher security areas where you can have a specific set of privilages and
+  still be logged in.
+
+  ### Error handler
+
+  When using plugs, you'll need to specify an error handler
+
+  The error_handler requires an `auth_error` function that receives the conn
+  and the error reason
+
+  ### Inline pipelines
+
+  If you want to define your pipeline inline, you can do so by using
+  `Guardian.Plug.Pipeline` as a plug itself.
+  You _must_ supply the module and error handler inline if you do this.
+
+  ```elixir
+  plug Guardian.Plug.Pipeline, module: MyApp.Tokens,
+                          error_handler: MyApp.AuthErrorHandler
+  plug Guardian.VerifyHeader, realm: "Bearer"
+  ```
+
+  Inline pipelines are also good to change the error handler that you want to use.
+
+  Note that you must set the pipeline before using other guardian plugs.
+
+  ```elixir
+  # Use the MyApp.AuthErrorHandler for downstream Guardian plugs
+  plug Guardian.Plug.Pipeline, module: MyApp.Tokens,
+                          error_handler: MyApp.AuthErrorHandler
+  plug Guardian.VerifyHeader, realm: "Bearer"
+
+  # Now change out the error handler for plugs downstream of this one.
+  plug Guardian.Plug.Pipeline, error_handler: MyApp.SpecialAuthErrorHandler
+  ```
+  """
+
+  import Plug.Conn
+
+  @doc """
+  Create your very own `Guardian.Plug.Pipeline`
+
+  Using this macro will make your module into a plug builder.
+  It will provide your pipeline with the Guardian implementation module and error
+  handler so that it can be used within your pipeline and downstream.
+  """
+  defmacro __using__(opts \\ []) do
+    quote do
+      use Plug.Builder
+      import Guardian.Plug.Pipeline
+
+      def init(opts) do
+        otp_app = Keyword.get(opts, :otp_app) ||
+          Keyword.get(unquote(opts), :otp_app)
+        if !otp_app do
```

Comment:
This requires you provide an `:otp_app` but if you configure via the plug that is unnecessary.